controllers/front/OnePageCheckoutController.php

```diff
@@ -0,0 +1,432 @@
+        $invalidAddressIds = $addressValidator->validateCartAddresses($cart);
+
+        // Build the currently selected address' warning message (if relevant)
+        if (!$customer->isGuest() && !empty($invalidAddressIds)) {
+            $this->checkoutWarning['address'] = [
+                'id_address' => (int) reset($invalidAddressIds),
+                'exception' => $this->trans(
+                    'Your address is incomplete, please update it.',
+                    [],
+                    'Shop.Notifications.Error'
+                ),
+            ];
+        }
+
+        // Prevent check for guests
+        if ($customer->id) {
+            // Prepare all other addresses' warning messages (if relevant).
+            // These messages are displayed when changing the selected address.
+            $allInvalidAddressIds = $addressValidator->validateCustomerAddresses($customer, $this->context->language);
+            $this->checkoutWarning['invalid_addresses'] = $allInvalidAddressIds;
+        }
+
+        if (isset($data['checksum']) && $data['checksum'] === $this->cartChecksum->generateChecksum($cart)) {
+            $process->restorePersistedData($data);
+        }
+    }
+
+    public function displayAjaxselectDeliveryOption(): void
+    {
+        $cart = $this->cart_presenter->present(
+            $this->context->cart,
+            true
+        );
+
+        ob_end_clean();
+        header('Content-Type: application/json');
+        $this->ajaxRender(json_encode([
+            'preview' => $this->render('checkout/_partials/cart-summary', [
+                'cart' => $cart,
+                'static_token' => Tools::getToken(false),
+            ]),
+        ]));
+    }
+
+    public function displayAjaxCheckCartStillOrderable(): void
+    {
+        $responseData = [
+            'errors' => false,
+            'cartUrl' => '',
+        ];
+
+        if ($this->context->cart->isAllProductsInStock() !== true
+            || $this->context->cart->checkAllProductsAreStillAvailableInThisState() !== true
+            || $this->context->cart->checkAllProductsHaveMinimalQuantities() !== true) {
+            $responseData['errors'] = true;
+            $responseData['cartUrl'] = $this->context->link->getPageLink('cart', null, null, ['action' => 'show']);
+        }
+
+        header('Content-Type: application/json');
+        $this->ajaxRender(json_encode($responseData));
+    }
+
+    /**
+     * Assign template vars related to page content.
+     *
+     * @see FrontController::initContent()
+     */
+    public function initContent(): void
+    {
+        if (Configuration::isCatalogMode()) {
+            Tools::redirect('index.php');
+        }
+
+        $this->restorePersistedData($this->checkoutProcess);
+        $this->checkoutProcess->handleRequest(
+            Tools::getAllValues()
+        );
+
+        $presentedCart = $this->cart_presenter->present($this->context->cart, true);
```

Comment:
Already a second cart present I see during this review, every cart present kills performance, it should be done once per request.